presto-main/src/main/java/com/facebook/presto/tz/JdkBasedDateTimeZone.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tz;
+
+import org.joda.time.DateTimeZone;
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.zone.ZoneOffsetTransition;
+import java.time.zone.ZoneRules;
+import java.util.Objects;
+
+public class JdkBasedDateTimeZone
+        extends DateTimeZone
+{
+    private final ZoneRules zoneRules;
+
+    public JdkBasedDateTimeZone(String zoneId)
+    {
+        super(zoneId);
+        this.zoneRules = ZoneId.of(zoneId).getRules();
+    }
+
+    @Override
+    public String getNameKey(long instant)
+    {
+        // It is ok to return a dummy value here.
+        // getNameKey is used in 6 instances in Joda
+        // * `DateTimeZone.getShortName`. Assuming DefaultNameProvider is being used,
+        //   it then invokes `DefaultNameProvider.getShortName`, which in turn invokes
+        //   `DefaultNameProvider.getNameSet`, which ignores the passed in nameKey,
+        //   other than validating that it's not null.
+        // * `DateTimeZone.getName`. Same as above.
+        // * `CachedDateTimeZone.Info.getNameKey`. It is used to implement getNameKey.
+        // * `DateTimeZoneBuilder.PrecalculatedZone.getNameKey`. Same as above.
+        // * `DateTimeZoneBuilder.writeTo`. It only uses getNameKey when the object
+        //   is an instance of `FixedDateTimeZone`.
+        // * `ZoneInfoCompiler.test`. It's only invoked with built-in implementations
+        //   of `DateTimeZone`.
+
+        // Use an obviously incorrect and easily searchable value here
+        return "presto-name-key-not-provided";
+    }
+
+    @Override
+    public int getOffset(long instant)
+    {
+        return zoneRules.getOffset(Instant.ofEpochMilli(instant)).getTotalSeconds() * 1000;
```

Comment:
Looking at the code, `Instant.ofEpochMilli(instant)).getTotalSeconds()` evaluates to `Math.floorDiv(epochMilli, 1000)`.  I suggest we inline that to avoid the possible object creation and additional divisions/checks.

Reply:
It's ` zoneRules.getOffset(  Instant.ofEpochMilli(instant)  )  .getTotalSeconds() * 1000`

Reply:
Oh sorry, I read that wrong... saw the parenthesis in the wrong location.  The way you have is the only way.